### ISSUE TITLE
Database checkpoints

### ISSUE THREAD

**Issue by pfebrer96** [aiidateam/aiida-core#4518]
### Is your feature request related to a problem? Please describe

I'm trying to move to the use of aiida for projects where I need to handle lots of simulations and it's great. However, there's one thing that I don't like. I'm scared of experimenting, because I'm afraid I will create lots of nodes that I won't be able to delete and will stay in my database making it less clean.

### Describe the solution you'd like

I'd like that there would be an easy way of defining database checkpoints in aiida. In this way one could do:

![checkpoint flow](https://user-images.githubusercontent.com/42074085/97290650-2194a100-1849-11eb-8727-f95dc1b9bf56.png)

### Describe alternatives you've considered

I talked to @bosonie and he suggested to use a test and a production database. However this doesn't solve my problem, because I want to test my flows on the actual "production nodes".

Also, it's not just about testing new workflows. Maybe there's a point when I need to launch a bunch of simulations but 2 days later I realize that I set a parameter wrong. I think it's of little relevance for provenance that I made a mistake and had to rerun all those simulations.

Since pk's are incremental, this should be quite easy to implement, wouldn't it? The point of having labelled checkpoints instead of me writing down the pk of the "checkpoint node" is that the first one would be safer and more convenient :) 

**PS:** Checkpoints can have other advantages such as people being able to follow the history of how the db has been created if you allow descriptions/annotations to them. So you could, for example (just thinking out loud here), check if a given calculation was launched before/after a certain checkpoint and then understand better why that simulation was launched.

**Comment by greschd** [comment]
Interesting proposal! I definitely think this use case should be supported in some way.

I'm not _entirely_ sure however checkpoints are the right way to do it. My concern is that it only works if no other calculations / workflows are running at the same time as the experimenting. If anything that you would like to keep is produced during that time, there isn't really a good way to keep it.

As a potential alternative, maybe we could make it easier to "activate" a `Group` in a given interpreter, so that anything launched there will be added to it (or maybe that exists already)? A "normal" `verdi node delete` on the group content should then be sufficient to wipe the experiment - because of the graph traversal rules, anything produced as a descendant of things launched during that session would also be wiped.

**Comment by chrisjsewell** [comment]
why not just use the archiving feature? https://aiida.readthedocs.io/projects/aiida-core/en/latest/howto/share_data.html?highlight=archive#sharing-aiida-archives
(currently undergoing major improvements by me 😄)

**Comment by greschd** [comment]
Yeah, that would work also - I guess it depends on the size of your DB, and whether you know exactly what you'll need for the experiments.

If really the entire DB should be there, I feel groups + deleting is the more "natural" approach. But always good to have options 👌

**Comment by pfebrer96** [comment]
Thanks for the comments!

> My concern is that it only works if no other calculations / workflows are running at the same time as the experimenting.

Hmm that's true, I hadn't thought about it.

> As a potential alternative, maybe we could make it easier to "activate" a Group in a given interpreter, so that anything launched there will be added to it (or maybe that exists already)?

The "putting everything inside a group" approach sounds interesting! Maybe the definition of a `on_stored` callback can be used for this:

```python
from aiida.orm import Node

def mark_as_experimental(node):
    load_group("experimental").add_nodes(node)

Node.on_stored = mark_as_experimental
```
as it satisfies the condition of being interpreter specific. Although not sure if this would work with the daemon?

> as well as any links or mutable data.

This problem probably isn't solved by the group approach either, is it? I guess you would have to keep a log of the changes somehow. If you are concerned about this, you should also be concerned about deletion of nodes and so on. The critical part for me is that currently it is very hard to find and delete all the spare useless nodes that you have created. If someone is concerned about the rest of stuff, probably a database backup/clone (as proposed by @chrisjsewell) is the best option.

It would be cool to have a DB with git-like functionality, though :)

**One important thing**: what if a node that you included in the `"experimental"` group is used by a different interpreter on a calculation that is "not experimental"? It is crucial that the node in question is removed from the group, otherwise you will end up deleting nodes that weren't supposed to get deleted :scream:

Regarding the archive approach:

> You can create an archive, import it into a fresh profile, then play around with it as much as you want.

Are all computers and codes setup and ready to be used if you import a database?

**Comment by pfebrer96** [comment]
> The "links and other mutable data" are taken care of when deleting things: AiiDA has "traversal rules" for which nodes need to be included when deleting or exporting in order to end up with a consistent graph.

I thought that by "other mutable data" you meant things like labels, which are impossible to keep track of unless you keep a log, aren't they?

> This is, of course also to be considered if you have something that you want to keep which depends on an "experimental" result - when you get to deleting the experimental result, the other calculation will also be deleted.

I think this is what I pointed out here:

> One important thing: what if a node that you included in the "experimental" group is used by a different interpreter on a calculation that is "not experimental"? It is crucial that the node in question is removed from the group, otherwise you will end up deleting nodes that weren't supposed to get deleted

or do you mean something different?

Does it make sense/is it easy to impose the condition that only nodes within the group can be deleted? Then the engine would be smart enough to say: "I'm not deleting this node because it would affect nodes that are outside the group"

**Comment by greschd** [comment]
> I thought that by "other mutable data" you meant things like labels, which are impossible to keep track of unless you keep a log, aren't they?

Right, you could still change labels on the "older" part of the database. What I meant is that the DB is guaranteed to be in a consistent state by the deletion operation - if you were to just delete by PK, for example you could create a finished calculation but its outputs are suddenly gone.

> or do you mean something different?

Hm, maybe I misunderstood you there. But the problem then arises (as you pointed out) that the daemon is also a different interpreter. If you are sure that you'll want to delete _all_ the descendants of the experimental things, you can let the deletion process make sure these nodes are all captured, as long as they were launched from an interpreter that uses the group.

> Does it make sense/is it possible to impose the condition that only nodes within the group can be deleted? Then the engine would be smart enough to say: "I'm not deleting this node because it would affect nodes that are outside the group"

It's definitely possible to implement, but I'm not sure how useful it would be given the above.

**Comment by pfebrer96** [comment]
> But the problem then arises (as you pointed out) that the daemon is also a different interpreter. If you are sure that you'll want to delete all the descendants of the experimental things, you can let the deletion process make sure these nodes are all captured, as long as they were launched from an interpreter that uses the group.

> It's definitely possible to implement, but I'm not sure how useful it would be given the above.

Hmm can I launch two different daemons in the same profile? To have an experimental daemon and a production daemon :thinking:

**Comment by greschd** [comment]
No.. that separation happens at the profile level. Which brings us back to @chrisjsewell's proposal 😄

**Comment by pfebrer96** [comment]
Then, what is the answer to these two questions? :)

> Are all computers and codes setup and ready to be used if you import a database?
> 
> Also, can two databases be merged together? Because if not, I guess this would suffer from the same drawback that checkpoints. I.e. you are either experimenting or in production mode. Or at least you can not persist your experiments in case they go well but you have done some other calculations in your production db.

**Comment by greschd** [comment]
You can definitely merge databases (nodes are identified by their UUID, so there are no duplication issues). I'm not 100% sure about the computers + codes, maybe they are disabled on import. But worst case, you'll have to quickly `verdi {code,computer} duplicate` them.

**Comment by pfebrer96** [comment]
> No.. that separation happens at the profile level. Which brings us back to @chrisjsewell's proposal smile

I just realised that you can increment the number of workers for a single daemon and each worker runs its own interpreter. Isn't it very suitable for this goal?? Since you didn't mention it, I guess there are some issues/pitfalls.

**Comment by greschd** [comment]
Yeah, you don't really have control over which processes go to which worker..
